upstairs/src/lib.rs

```diff
@@ -2659,12 +2861,47 @@ impl Downstairs {
     /**
      * Enqueue a new downstairs request.
      */
-    fn enqueue(&mut self, io: DownstairsIO) {
+    async fn enqueue(
+        &mut self,
+        mut io: DownstairsIO,
+        ds_done_tx: mpsc::Sender<u64>,
+    ) {
+        let mut skipped = 0;
         for cid in 0..3 {
             assert_eq!(io.state[&cid], IOState::New);
-            self.io_state_count.incr(&IOState::New, cid);
+
+            let current = self.ds_state[cid as usize];
+            // If a downstairs is faulted, we can move that job directly
+            // to IOState::Skipped
+            match current {
+                DsState::Faulted | DsState::OnlineRepair => {
+                    io.state.insert(cid, IOState::Skipped);
+                    self.io_state_count.incr(&IOState::Skipped, cid);
+                    skipped += 1;
+                }
+                _ => {
+                    self.io_state_count.incr(&IOState::New, cid);
+                }
+            }
+        }
+
+        // Puts the IO onto the downstairs work queue.
+        let ds_id = io.ds_id;
+        self.ds_active.insert(ds_id, io);
+
+        // If we skipped all three jobs, then it's possible no downstairs
+        // task is around to tell ds_up_listen() that this job is done, so
+        // do that work here.
```

Comment:
This occurs when all three downstairs are Faulted, right, because the tasks will be killed?

Reply:
Correct, if no downstairs connection exists, then there is no way for IO to make progress through crucible.  This would be true even before this PR, just taking place in a different location.